### PR TITLE
[GF-3828] Detect Local Environment, Add Dismissible Notice

### PR DESCRIPTION
**Pull request by mike-day**
## Tasks

- [X] I have signed a [Contributor License Agreement (CLA)](https://github.com/studiopress/pattern-manager/wiki/For-Developers#contributor-license-agreement-cla) with WP Engine.

### Summary of changes

Users of Pattern Manager should be working with a local development environment. This PR adds a simple check for a local environment, then displays a dismissible notice if a non-local environment is detected and the user has not previously dismissed the notice.

The notice displays an external link to download Local WP: https://localwp.com/?modal=download.

### How to test

Using Local WP
1. Checkout the branch
2. Navigate to the main patterns view (`/wp-admin/admin.php?page=pattern-manager`)
3. The notice should **_NOT_** display since you are on a local environment

Using a WPE-Hosted Site
1. Checkout the branch
2. Create or use an existing site in my_wpe (https://my.wpengine.com/bulk_sites)
3. Install and activate the attached, zipped plugin: [pattern-manager.0.3.0.zip](https://github.com/studiopress/pattern-manager/files/12043046/pattern-manager.0.3.0.zip)
4. Navigate to the main patterns view (`/wp-admin/admin.php?page=pattern-manager`)
5. The notice should display
6. Dismiss the notice and refresh the page
7. The notice should no longer be displayed

### Notes & Screenshots

pm_local-error

### CODE REVIEW

File: wp-modules/api-data/api-data.php [hunk 1]
```diff
@@ -122,6 +134,26 @@ function delete_pattern( $request ) {
...
 }
 
+/**
+ * Updates the list of sites that should not show environment notifications.
+ *
+ * @param WP_REST_Request $request Full data about the request.
+ * @return WP_REST_Response
+ */
+function update_dismissed_sites( $request ) {
+	$dismissed_sites = array_merge( get_dismissed_sites(), (array) get_current_blog_id() );
```
Comment (mike-day) [inline wp-modules/api-data/api-data.php:144]: Storing `$dismissed_sites` as an array here so dismissal will still work as expected for multisite.

If we stored just the site id, a multisite admin would be overwriting any previously dismissed sites each time they dismissed the notice (as opposed to adding the dismissal to an array of previously dismissed sites).

File: wp-modules/app/js/src/types.ts [hunk 2]
```diff
@@ -52,7 +54,3 @@ export type QueriedCategories = {
 	label: string;
 	name: string;
 }[];
-
-export type Theme = {
-	name: string;
-};
```
Comment (mike-day) [inline wp-modules/app/js/src/types.ts:58]: Cleaning up an unused type from `VersionControlNotice`.

File: wp-modules/api-data/api-data.php (hunk 1 above)
Comment (dreamwhisper) [reply wp-modules/api-data/api-data.php:144]: Nice

File: wp-modules/get-environment/get-environment.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+/**
+ * Checks that the current environment is locally hosted.
+ *
+ * @param string $environment The environment type to check against.
+ * @return boolean
+ */
+function check_environment( $environment ) {
+	return wp_get_environment_type() === $environment;
+}
```
Comment (mike-day) [inline wp-modules/get-environment/get-environment.php:52]: I couldn't get a mock to work for `wp_get_environment_type()`, so we can just use a param for the environment to check against instead.

Comment (kienstra) [reply wp-modules/get-environment/get-environment.php:52]: That's fine, it doesn't look like there's a filter we can mock it with.

File: wp-modules/api-data/api-data.php (hunk 1 above)
Comment (mike-day) [reply wp-modules/api-data/api-data.php:144]: Also calling out usage of `get_current_blog_id()` as the site id.

I think it makes sense as a value that should not change, regardless of if the name of a site changes. But please check me if there is a more appropriate value to use!

File: wp-modules/app/js/src/components/App/index.tsx [hunk 4]
```diff
@@ -44,10 +48,18 @@ export default function App() {
...
-					<VersionControlNotice
-						isVisible={ versionControl.displayNotice }
-						handleDismiss={ versionControl.updateDismissedThemes }
-					/>
+					<div className="patternmanager-notice-container">
+						<EnvironmentNotice
+							isVisible={ environment.displayNotice }
+							handleDismiss={ environment.updateDismissedSites }
+						/>
+						<VersionControlNotice
```
Comment (kienstra) [inline wp-modules/app/js/src/components/App/index.tsx:56]: This looks good.

If we ever add a 3rd, let's add some abstraction for them.

But 2 separate ones is good.

File: wp-modules/app/js/src/components/Notices/EnvironmentNotice/test/index.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+			handleDismiss={ handleDismissSpy }
+		/>
+	);
+
+	// Dismiss the notice.
+	act( () => {
+		testRenderer.root.findByProps( { type: 'button' } ).props.onClick();
+	} );
+
+	expect( handleDismissSpy ).toHaveBeenCalled();
```
Comment (kienstra) [inline wp-modules/app/js/src/components/Notices/EnvironmentNotice/test/index.spec.tsx:44]: Good one!

File: wp-modules/get-environment/tests/GetEnvironmentTest.php [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+		$this->assertSame( $this->get_mock_dismissed_sites(), get_dismissed_sites() );
+	}
+
+	/**
+	 * Tests `check_environment_notice_should_show()` from the `get-environment` module.
+	 */
+	public function test_check_environment_notice_should_show() {
+		$this->assertFalse( (bool) check_environment_notice_should_show( $this->get_mock_dismissed_sites()[0] ) );
+		$this->assertFalse( (bool) check_environment_notice_should_show( $this->get_mock_non_dismissed_site() ) );
+		$this->assertTrue( (bool) check_environment_notice_should_show( $this->get_mock_non_dismissed_site(), $this->get_mock_environment_type_hosted() ) );
```
Comment (kienstra) [inline wp-modules/get-environment/tests/GetEnvironmentTest.php:70]: This looks good.

Unit tests should never run in production, so `$this->get_mock_environment_type_hosted()` should make `check_environment_notice_should_show()` be `true`.

**Review by kienstra — APPROVED** [review]
Hi @mike-day,
Really nice work. 

Great unit test coverage, and this works as expected:

Screenshot 2023-07-13 at 12 55 25 PM

File: wp-modules/app/js/src/components/Notices/EnvironmentNotice/test/index.spec.tsx (hunk 5 above)
Comment (kienstra) [reply wp-modules/app/js/src/components/Notices/EnvironmentNotice/test/index.spec.tsx:44]: It only tests the API of `<EnvironmentNotice>`, that the handler is called. 

What the handler does isn't tested 👍

**Review by dreamwhisper — APPROVED** [review]
Nice work!

File: wp-modules/get-environment/get-environment.php [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+/**
+ * Determines if the environment notice should be displayed.
+ *
+ * @param string $site_id The site id to check against previously dismissed sites.
+ * @param string $environment The environment type to check against.
+ * @return boolean
+ */
+function check_environment_notice_should_show( $site_id, $environment = 'local' ) {
```
Comment (kienstra) [inline wp-modules/get-environment/get-environment.php:40]: Nice 2nd parameter, a sort of dependency injection to allow mocking.

File: wp-modules/app/js/src/types.ts (hunk 2 above)
Comment (kienstra) [reply wp-modules/app/js/src/types.ts:58]: Good idea!

File: wp-modules/app/js/src/hooks/useEnvironment.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
+import { useState } from '@wordpress/element';
+import { patternManager } from '../globals';
+import getHeaders from '../utils/getHeaders';
+
+export default function useEnvironment( initialEnvironmentNotice: boolean ) {
+	const [ displayNotice, setDisplayNotice ] = useState(
```
Comment (kienstra) [inline wp-modules/app/js/src/hooks/useEnvironment.ts:6]: Nice hook.

The state is very simple. Just `Boolean` state, and a handler to update it.

File: wp-modules/app/js/src/hooks/useEnvironment.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+		initialEnvironmentNotice
+	);
+
+	function updateDismissedSites() {
+		setDisplayNotice( false );
+		return fetch(
```
Comment (kienstra) [inline wp-modules/app/js/src/hooks/useEnvironment.ts:12]: It's nice how this `fetch()` is orthogonal to the `displayNotice` state.

There's no need to wait for the `fetch()` call to complete to update the `displayNotice` state.

File: wp-modules/app/js/src/components/App/index.tsx (hunk 4 above)
Comment (mike-day) [reply wp-modules/app/js/src/components/App/index.tsx:56]: Sounds like a plan!

File: wp-modules/app/js/src/hooks/useEnvironment.ts (hunk 9 above)
Comment (mike-day) [reply wp-modules/app/js/src/hooks/useEnvironment.ts:12]: Learned that one from you! Optimistic UI for the win 😊

Comment (kienstra) [reply wp-modules/app/js/src/hooks/useEnvironment.ts:12]: Haha, thanks!

**Pull request merged**